Lib/idlelib/idle_test/test_pyparse.py

```diff
@@ -0,0 +1,521 @@
+            TestInfo('":\n', no),
+            TestInfo('\n   def function1(self, a,\n', no),
+            TestInfo('def function1(self, a):\n    pass\n', no),
+            TestInfo('# A comment:\n', no),
+            TestInfo('"""A docstring:\n', no),
+            TestInfo('"""A docstring:\n', no),
+            )
+
+        for test in tests:
+            with self.subTest(string=test.string):
+                setstr(test.string)
+                test.assert_(opener())
+
+    def test_is_block_closer(self):
+        yes = self.assertTrue
+        no = self.assertFalse
+        p = self.parser
+        setstr = p.set_str
+        closer = p.is_block_closer
+
+        TestInfo = namedtuple('TestInfo', ['string', 'assert_'])
+        tests = (
+            TestInfo('return\n', yes),
+            TestInfo('\tbreak\n', yes),
+            TestInfo('  continue\n', yes),
+            TestInfo('     raise\n', yes),
+            TestInfo('pass    \n', yes),
+            TestInfo('pass\t\n', yes),
+            TestInfo('return #\n', yes),
+            TestInfo('raised\n', no),
+            TestInfo('returning\n', no),
+            TestInfo('# return\n', no),
+            TestInfo('"""break\n', no),
+            TestInfo('"continue\n', no),
+            TestInfo('def function1(self, a):\n    pass\n', yes),
+            )
+
+        for test in tests:
+            with self.subTest(string=test.string):
+                setstr(test.string)
+                test.assert_(closer())
+
+    def test_get_last_open_bracket_pos(self):
+        eq = self.assertEqual
+        p = self.parser
+        setstr = p.set_str
+        openbracket = p.get_last_open_bracket_pos
+
+        TestInfo = namedtuple('TestInfo', ['string', 'position'])
+        tests = (
+            TestInfo('', None),
+            TestInfo('a\n', None),
+            TestInfo('# (\n', None),
+            TestInfo('""" (\n', None),
+            TestInfo('a = (1 + 2) - 5 *\\\n', None),
+            TestInfo('\n   def function1(self, a,\n', 17),
+            TestInfo('\n   def function1(self, a,  # End of line comment.\n', 17),
+            TestInfo('{)(]\n', None),
+            TestInfo('(((((((((()))))))\n', 2),
+            TestInfo('(((((((((())\n)))\n))\n', 2),
+            )
+
+        for test in tests:
+            # There is a bug where the value is carried forward from last item.
+            p.lastopenbracketpos = None
+            with self.subTest(string=test.string):
+                setstr(test.string)
+                eq(openbracket(), test.position)
+
+    def test_get_last_stmt_bracketing(self):
+        eq = self.assertEqual
+        p = self.parser
+        setstr = p.set_str
+        bracketing = p.get_last_stmt_bracketing
+
+        TestInfo = namedtuple('TestInfo', ['string', 'bracket'])
+        tests = (
+            TestInfo('', ((0, 0),)),
+            TestInfo('a\n', ((0, 0),)),
+            TestInfo('()()\n', ((0, 0), (0, 1), (2, 0), (2, 1), (4, 0))),
```

Comment:
I understand that the code produces 4 rather than 3 but I wonder if the caller really cares.
These really test _study2 more, but with cases where we only care about the bracketing.

Reply:
I actually intended this test to show the level (second element) increase and decrease on closers.  I didn't look at how it was used, but just tried to make tests for what it was returning.